ray-operator/controllers/ray/rayjob_controller.go

```diff
@@ -133,30 +133,12 @@ func (r *RayJobReconciler) Reconcile(ctx context.Context, request ctrl.Request)
 		return ctrl.Result{RequeueAfter: RayJobDefaultRequeueDuration}, err
 	}
 
-	if err := utils.ValidateRayJobMetadata(rayJobInstance.ObjectMeta); err != nil {
-		logger.Error(err, "The RayJob metadata is invalid")
-		r.Recorder.Eventf(rayJobInstance, corev1.EventTypeWarning, string(utils.InvalidRayJobMetadata),
-			"The RayJob metadata is invalid %s/%s: %v", rayJobInstance.Namespace, rayJobInstance.Name, err)
-		return ctrl.Result{RequeueAfter: RayJobDefaultRequeueDuration}, err
-	}
-
-	if err := utils.ValidateRayJobSpec(rayJobInstance); err != nil {
-		logger.Error(err, "The RayJob spec is invalid")
-		r.Recorder.Eventf(rayJobInstance, corev1.EventTypeWarning, string(utils.InvalidRayJobSpec),
-			"The RayJob spec is invalid %s/%s: %v", rayJobInstance.Namespace, rayJobInstance.Name, err)
-		return ctrl.Result{RequeueAfter: RayJobDefaultRequeueDuration}, err
-	}
-
-	if err := utils.ValidateRayJobStatus(rayJobInstance); err != nil {
-		logger.Error(err, "The RayJob status is invalid")
-		r.Recorder.Eventf(rayJobInstance, corev1.EventTypeWarning, string(utils.InvalidRayJobStatus),
-			"The RayJob status is invalid %s/%s: %v", rayJobInstance.Namespace, rayJobInstance.Name, err)
-		return ctrl.Result{RequeueAfter: RayJobDefaultRequeueDuration}, err
-	}
-
 	// Please do NOT modify `originalRayJobInstance` in the following code.
 	originalRayJobInstance := rayJobInstance.DeepCopy()
 
+	// Perform all validations and directly fail the RayJob if any of the validation fails
+	validateErr := r.validateRayJobAndUpdateStatusIfNeeded(ctx, rayJobInstance)
+
```

Comment:
How about calling `updateRayJobStatus` here directly so that we don't need to go into the `switch` statement? This makes the implementation simpler.

The comment "This is the only place where we update the RayJob status." is a good suggestion, as it helps avoid calling `Status().Update()` inside the `switch` statement, where complexity typically increases significantly. However, since validation is not in the `switch` statement, updating the status immediately after validation is acceptable and actually reduces complexity.

https://github.com/ray-project/kuberay/blob/f18b1a48befa1258248b07b55d84a8d39a47e5f7/ray-operator/controllers/ray/rayjob_controller.go#L459-L461